### PR TITLE
Colorbar Update

### PR DESCRIPTION
**Pull request by bollu**
Port `ColorBarVisual`to the new SceneGraph API, along with test cases and examples.

### CODE REVIEW

**Comment by bollu** [comment]
@campagnola the `canvas` coordinate system corresponds to a correctly scaled coordinate system with a 1:1 relation with pixel size, correct?

**Comment by bollu** [comment]
I'm having strange behaviour with the border rendering in `Colorbar`

**Comment by campagnola** [comment]
Technically you should use the `document` cs for pixel measurements. In most cases, this will be the same as `canvas`, though.

**Comment by bollu** [comment]
since `document` is guaranteed to be isotropic, simply scaling the sides would work for the borders now I think? This would not have worked previously since the coordinate system was anisotropic

**Comment by Eric89GXL** [comment]
You should be able to do whatever is done in the second visual tutorial here, no?

https://github.com/campagnola/vispy/blob/scenegraph-update/examples/tutorial/visuals/T02_measurements.py

**Comment by bollu** [comment]
Yep, But I was curious if we could improve on that

**Comment by Eric89GXL** [comment]
Perhaps -- I think for now, though, we should go with whatever works / the lowest barrier to entry, making the highest priority getting the PR working so we can merge it.

**Comment by campagnola** [comment]
I think it's safe to make the assumption that colorbars are unscaled and unrotated with respect to their document (or even if they are, then we don't need to worry too much about how the border performs). I would use `Widget`'s approach to drawing borders in this case.

**Comment by campagnola** [comment]
What's the status of this? The `colorbar_visual_types` example doesn't look quite right:
titlepage1

**Comment by Eric89GXL** [comment]
@campagnola I think this is one of the final blockers for merge, right? If so, feel free to take over if it helps get the PR ready.

**Comment by bollu** [comment]
@campagnola yep, that was one of the problems I was working on. Sorry for not being around, had college work to complete :) I'm working on this tonight

**Comment by campagnola** [comment]
Looks good! Please check flake, and this will be ready to go.
Thanks for all your help getting this branch up to speed! It helped immensely in getting the code to a presentable state (and it was very well received at the SciPy conference last week).

File: examples/basics/visuals/colorbar_visual_types.py [hunk 1]
```diff
@@ -11,8 +11,10 @@
 from vispy import gloo
 from vispy.visuals import ColorBarVisual
 from vispy.color import get_colormap
-from vispy.visuals.transforms import NullTransform
-from vispy.visuals.transforms.transform_system import TransformSystem
+from vispy.color import Color
```
Comment (Eric89GXL) [inline examples/basics/visuals/colorbar_visual_types.py]: Combine with `vispy.color` import line above

File: examples/basics/visuals/colorbar_visual_types.py [hunk 2]
```diff
@@ -90,18 +92,26 @@ def __init__(self):
 
         # construct a default transform that does identity scaling
         # and does not translate the coordinates
-        transform = NullTransform()
-        self.transform_sys = TransformSystem(self)
-        self.transform_sys.visual_to_document = transform
+        # transform = NullTransform()
+        # self.transform_sys = TransformSystem(self)
+        # self.transform_sys.visual_to_document = transform
```
Comment (Eric89GXL) [inline examples/basics/visuals/colorbar_visual_types.py]: might as well just delete these, I don't think we'll need them anymore

File: vispy/visuals/colorbar.py [hunk 3]
```diff
@@ -409,33 +557,414 @@ def border_width(self, border_width):
...
+#         """
+
+#         self._program.draw('triangles')
+#         self.set_gl_state(cull_face=False)
+#         self._border_program.draw("triangle_strip")
+
+#         self._label.draw()
+
+#         for tick in self._ticks:
+#             tick.draw()
```
Comment (Eric89GXL) [inline vispy/visuals/colorbar.py]: I think you can just delete all of these, they are in the `git` history if we ever need them

**Comment by Eric89GXL** [comment]
Other than my minor comments, LGTM

**Comment by bollu** [comment]
@Eric89GXL there's still a bug - The borders don't render right. I'm not sure, but the transforms seem to be messed up.

@campagnola  I split the colorbar into a `CompoundVisual`, pushing the actual colorbar rendering into a subclass. Is this okay?

**Comment by Eric89GXL** [comment]
That's strange about the borders. Other visuals have borders that render okay -- can you find any differences between those and Colorbar?

I think it makes sense to split colorbar into a CompoundVisual with private Visual class that it uses.

**Comment by bollu** [comment]
borders work now. There was a bug with the way VBO's were sent :)

**Comment by Eric89GXL** [comment]
Great! So this is good to go then? Flake okay and tests pass?

**Comment by bollu** [comment]
will check testcases now. Flake I think has a few minor issues

**Comment by Eric89GXL** [comment]
On his PR, the last couple visuals mentioned are `Isoline` and `SurfacePlot`. I think someone got `IsoLine` recently, but can you check? And do `SurfacePlot` if it's not working yet? Then it should just be any remaining bugfixes at Luke's end, I think.

**Comment by bollu** [comment]
Yeah, IsoLine is done.

SurfacePlot "works" (as in, it runs). but there's a lighting artifact - the specular highlights appear even when the eye is _perpendicular_ to a normal (the highlights appear on the edges of a surface) which is wrong. I tried debugging it for a while, but I wasn't able to figure it out. I was planning on taking a look once this is done.

Other than that, I think AGG is left to be implemented?

**Comment by Eric89GXL** [comment]
Ahh, right. Forgot about that. Do tests pass for SurfacePlot? I think SurfacePlot, Iso*, and ColorBar were the only failing tests I had a few days ago.

**Comment by bollu** [comment]
Um, I'll have to switch branches and check. Gimme a minute

**Comment by bollu** [comment]
It doesn't seem to have an associated test is `vispy/visuals/testing`. I think we need to write more testcases as well?

```
 ~/p/vispy   mesh-visual-phong $  ls vispy/visuals/tests/ -1u                                           Wed Jul 15 02:39:08 IST 2015
__pycache__/
test_colorbar.py
test_volume.py
test_spectrogram.py
test_text.py
test_sdf.py
test_regular_polygon.py
test_rectangle.py
test_markers.py
test_polygon.py
test_histogram.py
test_image.py
test_ellipse.py
test_collections.py
__init__.py
```

**Comment by bollu** [comment]
Also, weird behavior on `test_colorbar`: it passes sometimes and fails sometimes. In 10 runs, it passed 5 times and failed 5 times. Can you check this on your end?

**Comment by bollu** [comment]
Silly me - I didn't port the tests :) Works fine now

**Comment by bollu** [comment]
fixed the flake errors too. Looks good from my end

**Comment by Eric89GXL** [comment]
Great! On to AGG, then?

**Comment by bollu** [comment]
yep! I'm unsure as to how to begin on AGG, but I guess I'll start by reading through the code as usual

**Comment by Eric89GXL** [comment]
Yeah I think that's the way to go. Hopefully you can just sub into the GLSL
the right variables like other visuals

**Comment by campagnola** [comment]
>  I split the colorbar into a CompoundVisual, pushing the actual colorbar rendering into a subclass. Is this okay?

That's what I would have done.

**Comment by campagnola** [comment]
Hooray, all visuals ported! 
Many thanks, @bollu!

**Pull request merged**